File: starlore-front/src/api/knowledgeMemory.ts
import request from '@/utils/request'

export interface KnowledgeArticleItem {
  articleId: number
  title: string
  description?: string
  category?: string
  cardCount: number
  generated: boolean
  updatedAt?: string
}

export interface KnowledgeGroupItem extends KnowledgeArticleItem {
  dueCount: number
  masteredCount: number
}

export interface KnowledgeCardItem {
  id: number
  articleId: number
  articleTitle: string
  question: string
  answerMarkdown: string
  sourceLineStart: number
  sourceLineEnd: number
  sourceHash: string
  difficulty: 'basic' | 'core' | 'advanced'
  sortOrder: number
  status: string
  reviewStage: string
  intervalDays: number
  reviewCount: number
  lapseCount: number
  nextReviewAt?: string
  lastReviewedAt?: string
}

export interface GenerateKnowledgeResult {
  generated: { articleId: number; title: string; cardCount: number }[]
  errors: { articleId: number; title?: string; message: string }[]
}

export function getKnowledgeArticles() {
  return request.get<never, { data: KnowledgeArticleItem[] }>('/knowledge-memory/articles')
}

export function getKnowledgeGroups() {
  return request.get<never, { data: KnowledgeGroupItem[] }>('/knowledge-memory/groups')
}

export function generateKnowledgeCards(data: {
  articleIds: number[]
  model?: string
  maxCardsPerArticle?: number
}) {
  return request.post<never, { success: boolean; message: string; data: GenerateKnowledgeResult }>(
    '/knowledge-memory/generate',
    data,
  )
}

export function getKnowledgeCards(articleId: number, dueOnly = false) {
  return request.get<never, { data: KnowledgeCardItem[] }>(
    `/knowledge-memory/articles/${articleId}/cards`,
    { params: { dueOnly } },
  )
}

export function reviewKnowledgeCard(
  cardId: number,
  data: { rating: 'again' | 'hard' | 'good'; durationMs?: number },
) {
  return request.post<never, { success: boolean; data: KnowledgeCardItem }>(
    `/knowledge-memory/cards/${cardId}/review`,
    data,
  )
}

export function deleteKnowledgeCard(cardId: number) {
  return request.delete(`/knowledge-memory/cards/${cardId}`)
}

File: starlore-py/app/schemas/knowledge_memory.py
"""知识记忆 API Schema。"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class GenerateKnowledgeCardsRequest(BaseModel):
    articleIds: list[int] = Field(min_length=1, max_length=10)
    model: str = ""
    maxCardsPerArticle: int = Field(default=20, ge=3, le=40)


class UpdateKnowledgeCardRequest(BaseModel):
    question: str | None = Field(default=None, min_length=1, max_length=500)
    answerMarkdown: str | None = Field(default=None, min_length=1)
    difficulty: Literal["basic", "core", "advanced"] | None = None
    status: Literal["active", "archived"] | None = None


class ReviewKnowledgeCardRequest(BaseModel):
    rating: Literal["again", "hard", "good"]
    durationMs: int | None = Field(default=None, ge=0, le=3600000)


class KnowledgeArticleItem(BaseModel):
    articleId: int
    title: str
    description: str | None = None
    category: str | None = None
    cardCount: int = 0
    generated: bool = False
    updatedAt: datetime | None = None


class KnowledgeGroupItem(KnowledgeArticleItem):
    dueCount: int = 0
    masteredCount: int = 0


class KnowledgeCardItem(BaseModel):
    id: int
    articleId: int
    articleTitle: str
    question: str
    answerMarkdown: str
    sourceLineStart: int
    sourceLineEnd: int
    sourceHash: str
    difficulty: str
    sortOrder: int
    status: str
    reviewStage: str
    intervalDays: int
    reviewCount: int
    lapseCount: int
    nextReviewAt: datetime | None = None
    lastReviewedAt: datetime | None = None
    createdAt: datetime | None = None
    updatedAt: datetime | None = None
